src/preprocess/index.ts

```diff
@@ -631,6 +631,47 @@ function validateTemplateParameter(param: $param, mergedParams: any, name: strin
     if (!(typeof paramValue === 'string' && paramValue.match(patternRegex))) {
       throw new Error(`Invalid value "${param.Name}" in ${name}. AllowedPattern: ${param.AllowedPattern}.`)
     }
+  } else if (typeof param.Type === 'string') {
+    const throwParamTypeError = () => {
+      throw new Error(`Invalid parameter value "${JSON.stringify(paramValue, null, ' ')}". Expected a ${param.Type}`);
+    };
+    switch (param.Type) {
+    case 'string':
+    case 'String':
+      if (!_.isString(paramValue)) {
+        throwParamTypeError();
+      }
+      break;
+    case 'number':
+    case 'Number':
+      if (!_.isNumber(paramValue)) {
+        throwParamTypeError();
+      }
+      break;
+    case 'object':
+    case 'Object':
+      if (!_.isObject(paramValue)) {
+        throwParamTypeError();
+      }
+      break;
+    case 'CommaDelimitedList':
+    case 'List<Number>':
+    case 'AWS::EC2::KeyPair::KeyName':
+    case 'AWS::EC2::SecurityGroup::Id':
+    case 'AWS::EC2::Subnet::Id':
+    case 'AWS::EC2::VPC::Id':
+    case 'List<AWS::EC2::VPC::Id>':
+    case 'List<AWS::EC2::SecurityGroup::Id>':
+    case 'List<AWS::EC2::Subnet::Id>':
```

Comment:
This might as well be an exhaustive list:
 
```
AWS::EC2::AvailabilityZone::Name
AWS::EC2::Image::Id
AWS::EC2::Instance::Id
AWS::EC2::KeyPair::KeyName
AWS::EC2::SecurityGroup::GroupName
AWS::EC2::SecurityGroup::Id
AWS::EC2::Subnet::Id
AWS::EC2::Volume::Id
AWS::EC2::VPC::Id
AWS::Route53::HostedZone::Id
List<AWS::EC2::AvailabilityZone::Name>
List<AWS::EC2::Image::Id>
List<AWS::EC2::Instance::Id>
List<AWS::EC2::SecurityGroup::GroupName>
List<AWS::EC2::SecurityGroup::Id>
List<AWS::EC2::Subnet::Id>
List<AWS::EC2::Volume::Id>
List<AWS::EC2::VPC::Id>
List<AWS::Route53::HostedZone::Id>
AWS::SSM::Parameter::Name
AWS::SSM::Parameter::Value<String>
AWS::SSM::Parameter::Value<List<String>>
AWS::SSM::Parameter::Value<CommaDelimitedList>
```

Although `AWS::SSM::Parameter::Value<AWS-specific parameter type>` and `AWS::SSM::Parameter::Value<List<AWS-specific parameter type>>` will take a bit of work to create.